From: David Woodhouse <dwmw@amazon.co.uk>
Date: Tue, 23 Jul 2024 23:37:08 +0100
Subject: dma: Automatically enable page touching on Caspian

On systems with memory overcommit, although no more pages will be *removed*
without the consent of virtio-balloon, not all pages are guaranteed to be
present at boot time. They need to be faulted in when mapped for DMA.

Enable this by default for the relevant EC2 instance types to avoid a lot
of time wasted in building separate images with the command line option.

Signed-off-by: David Woodhouse <dwmw@amazon.co.uk>
---
 kernel/dma/page_touching.c | 21 +++++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/kernel/dma/page_touching.c b/kernel/dma/page_touching.c
--- a/kernel/dma/page_touching.c
+++ b/kernel/dma/page_touching.c
@@ -34,6 +34,7 @@
 #include <linux/dma-map-ops.h>
 #include "direct.h"
 #include <linux/moduleparam.h>
+#include <linux/dmi.h>
 
 /*
  * A wrapper around dma_direct which does a readb on the memory being mapped
@@ -132,3 +133,23 @@ void setup_dma_page_touching_ops(struct device *dev)
 	dev_info(dev, "binding to page touching DMA ops\n");
 	dev->dma_ops = &page_touching_dma_ops;
 }
+
+static const struct dmi_system_id pt_enable_table[] __initconst = {
+	{
+		.matches = {
+			DMI_EXACT_MATCH(DMI_SYS_VENDOR, "Amazon EC2"),
+			DMI_MATCH(DMI_PRODUCT_NAME, "caspian"),
+		},
+	},
+	{},
+};
+
+static int __init dmi_enable_pt(void)
+{
+	if (dmi_check_system(pt_enable_table)) {
+		pr_info("Automatically enabling page touching for Caspian\n");
+		dma_page_touching_enable = 1;
+	}
+	return 0;
+}
+arch_initcall(dmi_enable_pt)

